File: code/packages/fsharp/aes/AesBlock.fs
namespace CodingAdventures.Aes.FSharp

open System

[<RequireQualifiedAccess>]
module AesBlock =
    let private blockSizeBytes = 16

    let private rcon =
        [| 0x00uy; 0x01uy; 0x02uy; 0x04uy; 0x08uy; 0x10uy; 0x20uy; 0x40uy; 0x80uy; 0x1buy; 0x36uy; 0x6cuy; 0xd8uy; 0xabuy; 0x4duy |]

    let private multiply (left: byte) (right: byte) =
        let mutable result = 0
        let mutable a = left
        let mutable b = right

        for _ in 0 .. 7 do
            if (b &&& 1uy) <> 0uy then
                result <- result ^^^ int a

            let carry = (a &&& 0x80uy) <> 0uy
            a <- byte ((int a <<< 1) &&& 0xff)
            if carry then
                a <- a ^^^ 0x1buy

            b <- b >>> 1

        byte result

    let private rotateLeft (value: byte) count =
        byte (((int value <<< count) ||| (int value >>> (8 - count))) &&& 0xff)

    let private multiplicativeInverse value =
        if value = 0uy then
            0uy
        else
            let mutable result = 1uy

            for _ in 0 .. 253 do
                result <- multiply result value

            result

    let private affineTransform value =
        value
        ^^^ rotateLeft value 1
        ^^^ rotateLeft value 2
        ^^^ rotateLeft value 3
        ^^^ rotateLeft value 4
        ^^^ 0x63uy

    let private buildSBoxes () =
        let sbox = Array.zeroCreate<byte> 256
        let invSbox = Array.zeroCreate<byte> 256

        for value in 0 .. 255 do
            let substituted = affineTransform (multiplicativeInverse (byte value))
            sbox[value] <- substituted
            invSbox[int substituted] <- byte value

        sbox, invSbox

    let private sbox, invSbox = buildSBoxes ()

    let private validateBlock (block: byte array) =
        if isNull block then nullArg "block"
        if block.Length <> blockSizeBytes then
            invalidArg "block" "AES block must be exactly 16 bytes."

    let private validateKey (key: byte array) =
        if isNull key then nullArg "key"
        if key.Length <> 16 && key.Length <> 24 && key.Length <> 32 then
            invalidArg "key" "AES key must be 16, 24, or 32 bytes."

    let private rotWord (word: byte array) =
        let first = word[0]
        word[0] <- word[1]
        word[1] <- word[2]
        word[2] <- word[3]
        word[3] <- first

    let private subWord (word: byte array) =
        for index in 0 .. word.Length - 1 do
            word[index] <- sbox[int word[index]]

    let private roundsForKeyLength keyLength =
        match keyLength with
        | 16 -> 10
        | 24 -> 12
        | 32 -> 14
        | _ -> invalidArg "key" "AES key must be 16, 24, or 32 bytes."

    let private expandKey (key: byte array) =
        let rounds = roundsForKeyLength key.Length
        let expanded = Array.zeroCreate<byte> (blockSizeBytes * (rounds + 1))
        Buffer.BlockCopy(key, 0, expanded, 0, key.Length)
        let temp = Array.zeroCreate<byte> 4
        let mutable bytesGenerated = key.Length
        let mutable rconIndex = 1

        while bytesGenerated < expanded.Length do
            for index in 0 .. 3 do
                temp[index] <- expanded[bytesGenerated - 4 + index]

            if bytesGenerated % key.Length = 0 then
                rotWord temp
                subWord temp
                temp[0] <- temp[0] ^^^ rcon[rconIndex]
                rconIndex <- rconIndex + 1
            elif key.Length = 32 && bytesGenerated % key.Length = 16 then
                subWord temp

            for index in 0 .. 3 do
                expanded[bytesGenerated] <- expanded[bytesGenerated - key.Length] ^^^ temp[index]
                bytesGenerated <- bytesGenerated + 1

        expanded

    let private addRoundKey (state: byte array) (expandedKey: byte array) round =
        let offset = round * blockSizeBytes

        for index in 0 .. blockSizeBytes - 1 do
            state[index] <- state[index] ^^^ expandedKey[offset + index]

    let private subBytes (state: byte array) =
        for index in 0 .. state.Length - 1 do
            state[index] <- sbox[int state[index]]

    let private invSubBytes (state: byte array) =
        for index in 0 .. state.Length - 1 do
            state[index] <- invSbox[int state[index]]

    let private shiftRows (state: byte array) =
        let copy = Array.copy state

        for row in 0 .. 3 do
            for column in 0 .. 3 do
                state[row + 4 * column] <- copy[row + 4 * ((column + row) &&& 3)]

    let private invShiftRows (state: byte array) =
        let copy = Array.copy state

        for row in 0 .. 3 do
            for column in 0 .. 3 do
                state[row + 4 * column] <- copy[row + 4 * ((column - row + 4) &&& 3)]

    let private mixColumns (state: byte array) =
        for column in 0 .. 3 do
            let offset = 4 * column
            let s0 = state[offset]
            let s1 = state[offset + 1]
            let s2 = state[offset + 2]
            let s3 = state[offset + 3]

            state[offset] <- multiply 0x02uy s0 ^^^ multiply 0x03uy s1 ^^^ s2 ^^^ s3
            state[offset + 1] <- s0 ^^^ multiply 0x02uy s1 ^^^ multiply 0x03uy s2 ^^^ s3
            state[offset + 2] <- s0 ^^^ s1 ^^^ multiply 0x02uy s2 ^^^ multiply 0x03uy s3
            state[offset + 3] <- multiply 0x03uy s0 ^^^ s1 ^^^ s2 ^^^ multiply 0x02uy s3

    let private invMixColumns (state: byte array) =
        for column in 0 .. 3 do
            let offset = 4 * column
            let s0 = state[offset]
            let s1 = state[offset + 1]
            let s2 = state[offset + 2]
            let s3 = state[offset + 3]

            state[offset] <- multiply 0x0euy s0 ^^^ multiply 0x0buy s1 ^^^ multiply 0x0duy s2 ^^^ multiply 0x09uy s3
            state[offset + 1] <- multiply 0x09uy s0 ^^^ multiply 0x0euy s1 ^^^ multiply 0x0buy s2 ^^^ multiply 0x0duy s3
            state[offset + 2] <- multiply 0x0duy s0 ^^^ multiply 0x09uy s1 ^^^ multiply 0x0euy s2 ^^^ multiply 0x0buy s3
            state[offset + 3] <- multiply 0x0buy s0 ^^^ multiply 0x0duy s1 ^^^ multiply 0x09uy s2 ^^^ multiply 0x0euy s3

    let encryptBlock (block: byte array) (key: byte array) =
        validateBlock block
        validateKey key

        let expandedKey = expandKey key
        let rounds = expandedKey.Length / blockSizeBytes - 1
        let state = Array.copy block

        addRoundKey state expandedKey 0

        for round in 1 .. rounds - 1 do
            subBytes state
            shiftRows state
            mixColumns state
            addRoundKey state expandedKey round

        subBytes state
        shiftRows state
        addRoundKey state expandedKey rounds
        state

    let decryptBlock (block: byte array) (key: byte array) =
        validateBlock block
        validateKey key

        let expandedKey = expandKey key
        let rounds = expandedKey.Length / blockSizeBytes - 1
        let state = Array.copy block

        addRoundKey state expandedKey rounds

        for round in rounds - 1 .. -1 .. 1 do
            invShiftRows state
            invSubBytes state
            addRoundKey state expandedKey round
            invMixColumns state

        invShiftRows state
        invSubBytes state
        addRoundKey state expandedKey 0
        state

File: code/packages/csharp/aes/AesBlock.cs
namespace CodingAdventures.Aes;

/// <summary>
/// AES single-block encryption and decryption helpers implemented from FIPS 197.
/// </summary>
public static class AesBlock
{
    private const int BlockSizeBytes = 16;
    private static readonly byte[] Rcon = [0x00, 0x01, 0x02, 0x04, 0x08, 0x10, 0x20, 0x40, 0x80, 0x1b, 0x36, 0x6c, 0xd8, 0xab, 0x4d];
    private static readonly byte[] SBox;
    private static readonly byte[] InvSBox;

    static AesBlock()
    {
        (SBox, InvSBox) = BuildSBoxes();
    }

    /// <summary>Encrypt one 16-byte block with a 16-, 24-, or 32-byte AES key.</summary>
    public static byte[] EncryptBlock(byte[] block, byte[] key)
    {
        ValidateBlock(block);
        ValidateKey(key);

        var expandedKey = ExpandKey(key);
        var rounds = expandedKey.Length / BlockSizeBytes - 1;
        var state = block.ToArray();

        AddRoundKey(state, expandedKey, 0);

        for (var round = 1; round < rounds; round++)
        {
            SubBytes(state);
            ShiftRows(state);
            MixColumns(state);
            AddRoundKey(state, expandedKey, round);
        }

        SubBytes(state);
        ShiftRows(state);
        AddRoundKey(state, expandedKey, rounds);
        return state;
    }

    /// <summary>Decrypt one 16-byte block with a 16-, 24-, or 32-byte AES key.</summary>
    public static byte[] DecryptBlock(byte[] block, byte[] key)
    {
        ValidateBlock(block);
        ValidateKey(key);

        var expandedKey = ExpandKey(key);
        var rounds = expandedKey.Length / BlockSizeBytes - 1;
        var state = block.ToArray();

        AddRoundKey(state, expandedKey, rounds);

        for (var round = rounds - 1; round >= 1; round--)
        {
            InvShiftRows(state);
            InvSubBytes(state);
            AddRoundKey(state, expandedKey, round);
            InvMixColumns(state);
        }

        InvShiftRows(state);
        InvSubBytes(state);
        AddRoundKey(state, expandedKey, 0);
        return state;
    }

    private static byte[] ExpandKey(byte[] key)
    {
        var keyWords = key.Length / 4;
        var rounds = keyWords switch
        {
            4 => 10,
            6 => 12,
            8 => 14,
            _ => throw new ArgumentException("AES key must be 16, 24, or 32 bytes.", nameof(key)),
        };

        var expanded = new byte[BlockSizeBytes * (rounds + 1)];
        Buffer.BlockCopy(key, 0, expanded, 0, key.Length);

        Span<byte> temp = stackalloc byte[4];
        var bytesGenerated = key.Length;
        var rconIndex = 1;

        while (bytesGenerated < expanded.Length)
        {
            for (var index = 0; index < 4; index++)
            {
                temp[index] = expanded[bytesGenerated - 4 + index];
            }

            if (bytesGenerated % key.Length == 0)
            {
                RotWord(temp);
                SubWord(temp);
                temp[0] ^= Rcon[rconIndex++];
            }
            else if (key.Length == 32 && bytesGenerated % key.Length == 16)
            {
                SubWord(temp);
            }

            for (var index = 0; index < 4; index++)
            {
                expanded[bytesGenerated] = (byte)(expanded[bytesGenerated - key.Length] ^ temp[index]);
                bytesGenerated++;
            }
        }

        return expanded;
    }

    private static void AddRoundKey(byte[] state, byte[] expandedKey, int round)
    {
        var offset = round * BlockSizeBytes;
        for (var index = 0; index < BlockSizeBytes; index++)
        {
            state[index] ^= expandedKey[offset + index];
        }
    }

    private static void SubBytes(byte[] state)
    {
        for (var index = 0; index < state.Length; index++)
        {
            state[index] = SBox[state[index]];
        }
    }

    private static void InvSubBytes(byte[] state)
    {
        for (var index = 0; index < state.Length; index++)
        {
            state[index] = InvSBox[state[index]];
        }
    }

    private static void ShiftRows(byte[] state)
    {
        var copy = state.ToArray();
        for (var row = 0; row < 4; row++)
        {
            for (var column = 0; column < 4; column++)
            {
                state[row + 4 * column] = copy[row + 4 * ((column + row) & 3)];
            }
        }
    }

    private static void InvShiftRows(byte[] state)
    {
        var copy = state.ToArray();
        for (var row = 0; row < 4; row++)
        {
            for (var column = 0; column < 4; column++)
            {
                state[row + 4 * column] = copy[row + 4 * ((column - row + 4) & 3)];
            }
        }
    }

    private static void MixColumns(byte[] state)
    {
        for (var column = 0; column < 4; column++)
        {
            var offset = 4 * column;
            var s0 = state[offset];
            var s1 = state[offset + 1];
            var s2 = state[offset + 2];
            var s3 = state[offset + 3];

            state[offset] = (byte)(Multiply(0x02, s0) ^ Multiply(0x03, s1) ^ s2 ^ s3);
            state[offset + 1] = (byte)(s0 ^ Multiply(0x02, s1) ^ Multiply(0x03, s2) ^ s3);
            state[offset + 2] = (byte)(s0 ^ s1 ^ Multiply(0x02, s2) ^ Multiply(0x03, s3));
            state[offset + 3] = (byte)(Multiply(0x03, s0) ^ s1 ^ s2 ^ Multiply(0x02, s3));
        }
    }

    private static void InvMixColumns(byte[] state)
    {
        for (var column = 0; column < 4; column++)
        {
            var offset = 4 * column;
            var s0 = state[offset];
            var s1 = state[offset + 1];
            var s2 = state[offset + 2];
            var s3 = state[offset + 3];

            state[offset] = (byte)(Multiply(0x0e, s0) ^ Multiply(0x0b, s1) ^ Multiply(0x0d, s2) ^ Multiply(0x09, s3));
            state[offset + 1] = (byte)(Multiply(0x09, s0) ^ Multiply(0x0e, s1) ^ Multiply(0x0b, s2) ^ Multiply(0x0d, s3));
            state[offset + 2] = (byte)(Multiply(0x0d, s0) ^ Multiply(0x09, s1) ^ Multiply(0x0e, s2) ^ Multiply(0x0b, s3));
            state[offset + 3] = (byte)(Multiply(0x0b, s0) ^ Multiply(0x0d, s1) ^ Multiply(0x09, s2) ^ Multiply(0x0e, s3));
        }
    }

    private static void RotWord(Span<byte> word)
    {
        var first = word[0];
        word[0] = word[1];
        word[1] = word[2];
        word[2] = word[3];
        word[3] = first;
    }

    private static void SubWord(Span<byte> word)
    {
        for (var index = 0; index < word.Length; index++)
        {
            word[index] = SBox[word[index]];
        }
    }

    private static (byte[] SBox, byte[] InvSBox) BuildSBoxes()
    {
        var sbox = new byte[256];
        var invSbox = new byte[256];

        for (var value = 0; value < 256; value++)
        {
            var substituted = AffineTransform(MultiplicativeInverse((byte)value));
            sbox[value] = substituted;
            invSbox[substituted] = (byte)value;
        }

        return (sbox, invSbox);
    }

    private static byte AffineTransform(byte value) =>
        (byte)(value ^ RotateLeft(value, 1) ^ RotateLeft(value, 2) ^ RotateLeft(value, 3) ^ RotateLeft(value, 4) ^ 0x63);

    private static byte MultiplicativeInverse(byte value)
    {
        if (value == 0)
        {
            return 0;
        }

        var result = (byte)1;
        for (var index = 0; index < 254; index++)
        {
            result = Multiply(result, value);
        }

        return result;
    }

    private static byte Multiply(byte left, byte right)
    {
        var result = 0;
        var a = left;
        var b = right;

        for (var index = 0; index < 8; index++)
        {
            if ((b & 1) != 0)
            {
                result ^= a;
            }

            var carry = (a & 0x80) != 0;
            a = (byte)(a << 1);
            if (carry)
            {
                a ^= 0x1b;
            }

            b >>= 1;
        }

        return (byte)result;
    }

    private static byte RotateLeft(byte value, int count) =>
        (byte)(((value << count) | (value >> (8 - count))) & 0xff);

    private static void ValidateBlock(byte[] block)
    {
        ArgumentNullException.ThrowIfNull(block);
        if (block.Length != BlockSizeBytes)
        {
            throw new ArgumentException("AES block must be exactly 16 bytes.", nameof(block));
        }
    }

    private static void ValidateKey(byte[] key)
    {
        ArgumentNullException.ThrowIfNull(key);
        if (key.Length is not (16 or 24 or 32))
        {
            throw new ArgumentException("AES key must be 16, 24, or 32 bytes.", nameof(key));
        }
    }
}
